src/SAML2/XML/saml/Attribute.php

```diff
@@ -129,65 +133,54 @@ public function getFriendlyName(): ?string
      * Set the value of the FriendlyName-property
      *
      * @param string|null $friendlyName
-     * @return void
      */
-    private function setFriendlyName(string $FriendlyName = null): void
+    private function setFriendlyName(?string $friendlyName): void
     {
-        $this->FriendlyName = $FriendlyName;
+        $this->FriendlyName = $friendlyName;
     }
 
 
     /**
      * Collect the value of the attributeValues-property
      *
-     * @return \SAML2\XML\saml\AttributeValue[]|null
+     * @return AttributeValue[]|null
      */
     public function getAttributeValues(): ?array
     {
-        return $this->attributeValues;
+        return $this->AttributeValues;
     }
 
 
     /**
      * Set the value of the AttributeValues-property
      *
      * @param array $attributeValues|null
-     * @return void
      */
-    private function setAttributeValues(?array $attributeValues): void
+    protected function setAttributeValues(?array $attributeValues): void
     {
-        $this->attributeValues = $attributeValues;
-    }
-
-
-    /**
-     * Add the value to the AttributeValues-property
-     *
-     * @param \SAML2\XML\saml\AttributeValue $attributeValue
-     * @return void
-     */
-    public function addAttributeValue(AttributeValue $attributeValue): void
-    {
-        $this->setAttributeValues(
-            empty($this->attributeValues) ? [$attributeValue] : array_merge($this->attributeValues, [$attributeValue])
-        );
+        if ($attributeValues === null) {
+            return;
+        }
+        Assert::allIsInstanceOf($attributeValues, AttributeValue::class, 'Invalid AttributeValue.');
+        $this->AttributeValues = $attributeValues;
     }
```

Comment:
Why are you turning an adder-method into a setter-method?

Reply:
I'm not, I just removed the adder, since I'm building the entire set manually and calling the setter. Having immutable objects, there's no point in having the adder, so I just removed it as in other places of the code.

Reply:
My bad..  BTW, I didn't notice you had merged this one into the other PR.. Is it still useful to keep this one open?

Reply:
I'll leave it up to you, as I don't know if you were planning to do any more work on this, or it works fine as it is now. Considering that there are merge conflicts, it might make sense to just drop it if there's nothing else we want to add to this one...